### PR TITLE
Basic SKU tracking for transport generation

### PR DESCRIPTION
**Pull request by luca-della-vedova**
## Summary

Automatically generate transport request by tracking input and output items of steps.
A new behavior tree node has been added to create a transport task if:

* The workcell has an `input item`.
* The `input item` is the `output item` of a workcell.
* The workcell that output the item is different from the workcell requesting it as an input.

Under these conditions, the system orchestrator will automatically generate and dispatch a transportation order to pickup the item from the workcell that output it and send it to the workcell that needs it.
Otherwise:

* If the requested input item is nowhere, the work order will fail (there is no record of where the item is).
* If the requested item is already in the same workcell, transportation will be silently skipped.

This means that work orders should start with a step that outputs the item that is required by the following steps, which I would argue is a fair additional requirement.

This new requirement breaks some tests and work orders but I would argue the breakage is good. For example, we test a work order that only [picks an item from a conveyor](https://github.com/osrf/nexus/blob/44315f87a42d0c76ae5868084a5465e00c7e80b5/nexus_demos/test_pick_and_place.py#L71). The work order however, [has an Input item](https://github.com/osrf/nexus/blob/44315f87a42d0c76ae5868084a5465e00c7e80b5/nexus_demos/config/pick_from_conveyor.json#L6-L14) that is not specified anywhere else, hence it is unclear how the workcell is supposed to retrieve that item.
By contrast, I changed the test to run the `place_on_conveyor` work order instead. This step has no input (hence no dependency) but has [one output](https://github.com/osrf/nexus/blob/44315f87a42d0c76ae5868084a5465e00c7e80b5/nexus_demos/config/place_on_conveyor.json#L6-L14), and succeeds correctly.

### Multi-pickup single dropoff tasks
Currently, `WorkcellTask`s have a single input and single output item, both are stored as a `string` with the name of  the item.
Moving them to be vector of items should be trivial but hasn't been done in this PR because there is no clear integration test to make sure the functionality works and there might be edge cases that we need to be careful about.
Specifically, multiple output items would just update the location of multiple items so that's trivial. Multiple input items could generate a "multi-pickup single dropoff" RMF tasks. This is trivial from a task generation point of view but  I fear there might be risk of deadlocks the moment we start creating a complex composed task, so I haven't done it yet.
We could also consider having an `Item` message but again not part of this PR because there is no need for it yet.

### Create a real dispenser
Since RMF delivery tasks are made of a pickup and dropoff, we need to have two different signaling paths:

* `dropoff` is fairly straightforward and is based on the existing implementation (I actually changed the hardcoded `pickup` into a hardcoded `dropoff` since that's how a delivery task will end).
* `pickup`, when the robot is sent to an arbitrary workcell it will send a dispenser request and the workcell should deliver its item on the robot is a lot trickier. It is unclear how this can be done right now (trigger a BT? have a workcell capability?). I added a `mock_dispenser_interface` to the RMF transporter plugin that will create a mock "always success" dispenser so we can test the rest of the integration first.

### More sophisticated SKU tracking

Currently, the location of SKUs is just tracked through a `std::unordered_map<std::string, std::string>` mapping SKU name to the name of the workcell that had it as an output item. This is done in the system orchestrator's `Context`.
Given that SKUs are (to my knowledge) globally unique identifiers we can have a single node manage all of them rather than do it locally in contexts. This node could offer ROS 2 interfaces to, for example, publish the state of all SKUs for visualization purposes or get the state of single SKUs.

### Multiple input / output stations per workcell

Workcells might have multiple input or output stations, for example an output into a conveyor and one into an AMR. Those would need different names which is not supported in this implementation yet.

### CODE REVIEW

File: nexus_demos/config/system_bts/pick_and_place.xml [hunk 1]
```diff
@@ -4,8 +4,10 @@
       <IsPauseTriggered paused="{paused}"/>
       <Parallel success_threshold="-1">
         <Sequence>
-          <BidTransporter name="bid_transporterasdasd" destination="{workcell}" result="{transporter}"/>
-          <TransporterRequest name="transporter_request" transporter="{transporter}" destination="{workcell}"/>
+          <!-- Create transport task, return success but a nullopt if it's not necessary -->
+          <CreateTransporterTask workcell_task="{task}" destination="{workcell}" result="{transport_task}"/>
+          <BidTransporter name="bid_transporterasdasd" transport_task="{transport_task}" result="{transporter}"/>
+          <TransporterRequest name="transporter_request" transport_task="{transport_task}" transporter="{transporter}"/>
```
Comment (Yadunund) [inline nexus_demos/config/system_bts/pick_and_place.xml:10]: Wrap in if-else https://github.com/BehaviorTree/BehaviorTree.CPP/blob/master/include/behaviortree_cpp/controls/if_then_else_node.h

Comment (luca-della-vedova) [reply nexus_demos/config/system_bts/pick_and_place.xml:10]: Note that this will most likely require an additional node.
The `IfThenElse` node can only deal with two conditions but currently, the `CreateTransporterTask` has three different states that it accounts for:

* FAILURE: The transportation task cannot be generated because, for example, the `InputItem` required by the workcell cannot be found anywhere in the facility. I believe this should fail the work order.
* NOOP (SUCCESS but return `std::nullopt` currently): The item is already in the workcell or there are no input items hence we can skip transportation and start the workcell straight away.
* TRANSPORT (SUCCESS with a filled `std::optional`): The item is in another workcell and needs to be transported.

I think there should still be three outputs and bundling `FAILURE` and `NOOP` together is not a great idea, so I suppose the only way to make this work would be to have something like:

```
<!-- Return the same as above, SUCCESS, and a std::optional -->
<CreateTransporterTask workcell_task="{task}" destination="{workcell}" result="{maybe_transport_task}"/>
<IfThenElse>
    <!-- Unpack the std::optional, return SUCCESS and its content in `output` if filled, FAILURE if not -->
    <UnpackTask input="{maybe_transport_task}" output="{transport_task}" />
    <Sequence>
        <BidTransporter name="bid_transporterasdasd" transport_task="{transport_task}" result="{transporter}"/>
        <TransporterRequest name="transporter_request" transport_task="{transport_task}" transporter="{transporter}"/>
    </Sequence>
</IfThenElse>
```

Comment (luca-della-vedova) [reply nexus_demos/config/system_bts/pick_and_place.xml]: This is what it would look like https://github.com/osrf/nexus/pull/101/commits/866ecd4bd3e2792c3e9b041519fbb6eb302dfb26

**Comment by Yadunund** [comment]
Feedback on some topics you highlighted in the description

> This new requirement breaks some tests and work orders but I would argue the breakage is good. For example, we test a work order that only [picks an item from a conveyor](https://github.com/osrf/nexus/blob/44315f87a42d0c76ae5868084a5465e00c7e80b5/nexus_demos/test_pick_and_place.py#L71). The work order however, [has an Input item](https://github.com/osrf/nexus/blob/44315f87a42d0c76ae5868084a5465e00c7e80b5/nexus_demos/config/pick_from_conveyor.json#L6-L14) that is not specified anywhere else, hence it is unclear how the workcell is supposed to retrieve that item.

Makes sense to me. 

> pickup, when the robot is sent to an arbitrary workcell it will send a dispenser request and the workcell should deliver its item on the robot is a lot trickier. It is unclear how this can be done right now (trigger a BT? have a workcell capability?). I added a mock_dispenser_interface to the RMF transporter plugin that will create a mock "always success" dispenser so we can test the rest of the integration first.

Sounds like we should add some APIs to the transporter base class to coordinate pickup & dropoff which transporters can implement? 

> Given that SKUs are (to my knowledge) globally unique identifiers we can have a single node manage all of them rather than do it locally in contexts. This node could offer ROS 2 interfaces to, for example, publish the state of all SKUs for visualization purposes or get the state of single SKUs.

Agreed. We should have a dedicated node that keeps track of products and their states.

File: nexus_system_orchestrator/src/workcell_request.cpp [hunk 2]
```diff
@@ -120,6 +120,11 @@ bool WorkcellRequest::on_result(
   }
 
   this->_ctx->set_task_results(result.result->result); // -.-
+  // Set output items of the step to be in this workcell
```
Comment (Yadunund) [inline nexus_system_orchestrator/src/workcell_request.cpp:123]: Wondering if this should be set after a bid is awarded and task is queued in a workcell in addition to after execution.

**Review by Yadunund — CHANGES_REQUESTED** [review]
Minor suggestions to consider.

File: nexus_system_orchestrator/src/context.cpp [hunk 3]
```diff
@@ -272,4 +272,22 @@ std::optional<std::vector<std::string>> Context::get_task_queued_signals(
   return it->second;
 }
 
+void Context::set_sku_location(const WorkcellTask& task, const std::string& workcell)
+{
+  if (task.output_item.size() > 0)
+  {
+    this->sku_locations[task.output_item] = workcell;
+  }
+}
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply nexus_system_orchestrator/src/context.cpp:281]: That's a good point, thanks for explaining. But yeah we can definitely make changes along the way if needed

File: nexus_system_orchestrator/src/workcell_request.cpp (hunk 2 above)
Comment (luca-della-vedova) [reply nexus_system_orchestrator/src/workcell_request.cpp]: I don't think it should be set when the task is queued. All the tasks are queued at the same time when the work order is submitted, and that doesn't really give us any information of where the item is.
What we could do is set it when the tasks start executing rather than when it ends executing.
I think this will make the most sense once we have input / output stations, then I imagine a SKU tracking workflow like:

* Transporter picks up item, SKU tracking node listens to transporters and sets the location to the transporter's location.
* Transporter drops off item at input station, SKU node sets the location to the input station.
* `WorkcellRequest` starts, SKU node sets the location to the workcell itself.
* `WorkcellRequest` ends, SKU node sets the location to the step's output station.

I added a TODO about this being the desired workflow (together with adding a new location update when the workcell starts processing the order) in https://github.com/osrf/nexus/pull/101/commits/9eb977451d40f257919dff03a3e7c3c2f0d30a8b

File: nexus_system_orchestrator/src/create_transporter_task.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  // Get the source as the location of the input SKU
+  if (workcell_task->input_item.size() == 0)
+  {
+    // We actually don't need to transport anything here, just return success
+    // and a nullopt transportation task
+    return BT::NodeStatus::SUCCESS;
+  }
+  const auto sku_position = this->_ctx->get_sku_location(workcell_task->input_item);
```
[the remark this answers is not in the document]
Comment (luca-della-vedova) [reply nexus_system_orchestrator/src/create_transporter_task.cpp]: https://github.com/osrf/nexus/pull/101/commits/560d3d4a972b0f9abc71c102ec759a202fae3d7f

File: nexus_system_orchestrator/src/create_transporter_task.hpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+#include "context.hpp"
+#include "session.hpp"
+
+#include <nexus_orchestrator_msgs/msg/workcell_task.hpp>
+#include <nexus_transporter_msgs/msg/transportation_request.hpp>
+
+#include <behaviortree_cpp_v3/action_node.h>
+
+#include <rclcpp_action/rclcpp_action.hpp>
+#include <rclcpp_lifecycle/lifecycle_node.hpp>
```
[the remark this answers is not in the document]
Comment (luca-della-vedova) [reply nexus_system_orchestrator/src/create_transporter_task.hpp]: https://github.com/osrf/nexus/pull/101/commits/332b37c875d1f1f4a0d3d5ad93181ff354450c51

File: nexus_transporter/src/transporters/rmf_transporter.cpp [hunk 6]
```diff
@@ -892,8 +956,8 @@ class RmfTransporter : public Transporter
     }
 
     // TODO(luca) also check for ingestors
```
[the remark this answers is not in the document]
Comment (luca-della-vedova) [reply nexus_transporter/src/transporters/rmf_transporter.cpp]: https://github.com/osrf/nexus/pull/101/commits/3e1157f1218dfba726e6a47d34b82fba68668bdc

File: nexus_system_orchestrator/src/create_transporter_task.cpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "create_transporter_task.hpp"
+
+#include <nexus_orchestrator_msgs/msg/workcell_task.hpp>
+#include <nexus_transporter_msgs/msg/destination.hpp>
```
[the remark this answers is not in the document]
Comment (Yadunund) [reply nexus_system_orchestrator/src/create_transporter_task.cpp:21]: I learn something new every day 😁

**Review by aaronchongth — APPROVED** [review]
I guess we can consider the flaky integration tests as a separate issue, and shouldn't block this. LGTM!
